passes/techmap/clockgate.cc

```diff
@@ -0,0 +1,230 @@
+		log("    -tie_lo <port_name>\n");
+		log("        Port <port_name> of the ICG will be tied to zero.\n");
+		log("        Intended for DFT scan-enable pins.\n");
+		log("    -min_net_size <n>\n");
+		log("        Only work on clocks with at least <n> eligible FFs.\n");
+		// log("        \n");
+	}
+
+	SigMap sigmap;
+	FfInitVals initvals;
+
+	// One ICG will be generated per ClkNetInfo
+	// if the number of FFs associated with it is sufficent
+	struct ClkNetInfo {
+		// Original, ungated clock into enabled FF
+		Wire* clk_net;
+		// Original clock enable into enabled FF
+		Wire* ce_net;
+		bool pol_clk;
+		bool pol_ce;
+		unsigned int hash() const {
+			auto t = std::make_tuple(clk_net, ce_net, pol_clk, pol_ce);
+			unsigned int h = mkhash_init;
+			h = mkhash(h, hash_ops<decltype(t)>::hash(t));
+			return h;
+		}
+		bool operator==(const ClkNetInfo& other) const {
+			return (clk_net == other.clk_net) &&
+			       (ce_net == other.ce_net) &&
+			       (pol_clk == other.pol_clk) &&
+			       (pol_ce == other.pol_ce);
+		}
+	};
+
+	struct GClkNetInfo {
+		// How many CE FFs on this CLK net have we seen?
+		int net_size;
+		// After ICG generation, we have new gated CLK signals
+		Wire* new_net;
+	};
+
+	ClkNetInfo clk_info_from_ff(FfData& ff) {
+		Wire* clk = ff.sig_clk.as_wire();
+		Wire* ce = ff.sig_ce.as_wire();
+		ClkNetInfo info{clk, ce, ff.pol_clk, ff.pol_ce};
+		return info;
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) override {
+		log_header(design, "Executing CLOCK_GATE pass (extract clock gating out of flip flops).\n");
+
+		std::optional<ClockGateCell> pos_icg_desc;
+		std::optional<ClockGateCell> neg_icg_desc;
+		std::vector<std::string> tie_lo_ports;
+		int min_net_size = 0;
+
+		size_t argidx;
+		for (argidx = 1; argidx < args.size(); argidx++) {
+			if (args[argidx] == "-pos" && argidx+2 < args.size()) {
+				auto name = args[++argidx];
+				auto rest = args[++argidx];
+				pos_icg_desc = icg_from_arg(name, rest);
+			}
+			if (args[argidx] == "-neg" && argidx+2 < args.size()) {
+				auto name = args[++argidx];
+				auto rest = args[++argidx];
+				neg_icg_desc = icg_from_arg(name, rest);
+			}
+			if (args[argidx] == "-tie_lo" && argidx+1 < args.size()) {
+				tie_lo_ports.push_back(RTLIL::escape_id(args[++argidx]));
+			}
+			if (args[argidx] == "-min_net_size" && argidx+1 < args.size()) {
+				min_net_size = atoi(args[++argidx].c_str());
+			}
+		}
+
+		extra_args(args, argidx, design);
+
+		pool<Cell*> ce_ffs;
+		dict<ClkNetInfo, GClkNetInfo> clk_nets;
+
+		int gated_flop_count = 0;
+		for (auto module : design->selected_whole_modules()) {
+			sigmap.set(module);
+			initvals.set(&sigmap, module);
+			for (auto cell : module->cells()) {
+				if (!RTLIL::builtin_ff_cell_types().count(cell->type))
+					continue;
+
+				FfData ff(&initvals, cell);
+				// It would be odd to get constants, but we better handle it
+				if (ff.has_ce && ff.sig_clk.is_wire() && ff.sig_ce.is_wire()) {
```

Comment:
This disqualifies cases where the enable signal is a single bit taken out of a multi-bit wire. See what `is_wire()` checks for